File: app/src/main/java/com/main/climbingdiary/view/TimeSlider.java
package com.main.climbingdiary.view;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.database.Cursor;
import android.text.TextUtils;
import android.util.Log;
import android.widget.TextView;

import com.crystal.crystalrangeseekbar.interfaces.OnRangeSeekbarChangeListener;
import com.crystal.crystalrangeseekbar.interfaces.OnRangeSeekbarFinalValueListener;
import com.crystal.crystalrangeseekbar.widgets.CrystalRangeSeekbar;
import com.main.climbingdiary.R;
import com.main.climbingdiary.view.button.ShowTimeSlider;
import com.main.climbingdiary.view.menu.MenuValues;
import com.main.climbingdiary.database.TaskRepository;
import com.main.climbingdiary.models.Filter;

import java.util.ArrayList;
import java.util.Collections;

public class TimeSlider implements OnRangeSeekbarChangeListener, OnRangeSeekbarFinalValueListener {
    private static CrystalRangeSeekbar rangeSeekbar;
    private static TextView minText;
    private static TextView maxText;
    private final static TaskRepository taskRepository = TaskRepository.getInstance();

    public TimeSlider(Activity _activity){
        rangeSeekbar = _activity.findViewById(R.id.timerange);
        minText = _activity.findViewById(R.id.textMin);
        maxText = _activity.findViewById(R.id.textMax);
        setTimes();
        rangeSeekbar.setOnRangeSeekbarChangeListener(this);
        rangeSeekbar.setOnRangeSeekbarFinalValueListener(this);
    }

    @SuppressLint("SetTextI18n")
    private static void setTimes(){
        try {
            ArrayList<Integer> years = new ArrayList<>();
            Cursor cursor = taskRepository.getYears();
            if (cursor != null) {
                while (!cursor.isAfterLast()) {
                    years.add(cursor.getInt(0));
                    cursor.moveToNext();
                }
            }
            Log.d("Years set", TextUtils.join(",", years));
            int minYear = Collections.min(years);
            int maxYear = Collections.max(years);
            Log.d("MIN || MAX", minYear + "||" + maxYear);
            rangeSeekbar.setMinValue(minYear);
            rangeSeekbar.setMaxValue(maxYear);
            rangeSeekbar.setMinStartValue(minYear);
            rangeSeekbar.setMaxStartValue(maxYear);
            minText.setText(Integer.toString(minYear));
            maxText.setText(Integer.toString(maxYear));
            ShowTimeSlider.showButton();
        }catch(Exception ex){
            ShowTimeSlider.hideButton();
        }
    }

    @Override
    public void valueChanged(Number minValue, Number maxValue) {
        String min = String.valueOf(minValue);
        String max = String.valueOf(maxValue);
        minText.setText(min);
        maxText.setText(max);
    }

    @Override
    public void finalValue(Number minValue, Number maxValue) {
        String min = String.valueOf(minValue);
        String max = String.valueOf(maxValue);
        if(!minValue.equals(maxValue)){
            String filter = "CAST(strftime('%Y',r.date) as int)>="+min+" and CAST(strftime('%Y',r.date) as int) <="+max;
            Filter.setFilter(filter, MenuValues.SORT_DATE);
        }else{
            String filter = "CAST(strftime('%Y',r.date) as int)=="+max;
            Filter.setFilter(filter,MenuValues.SORT_DATE);
        }
        Log.d("Filter set",Filter.getFilter());
        minText.setText(min);
        maxText.setText(max);
        FragmentPager.refreshAllFragments();
    }
}

File: app/src/main/java/com/main/climbingdiary/controller/slider/TimeSlider.kt
package com.main.climbingdiary.controller.slider

import android.text.TextUtils
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.google.android.material.slider.Slider
import com.google.android.material.slider.Slider.OnSliderTouchListener
import com.main.climbingdiary.R
import com.main.climbingdiary.activities.MainActivity
import com.main.climbingdiary.common.preferences.AppPreferenceManager
import com.main.climbingdiary.controller.FragmentPager.refreshSelectedFragment
import com.main.climbingdiary.controller.button.ShowTimeSlider.hideButton
import com.main.climbingdiary.database.TaskRepository.getYears
import com.main.climbingdiary.models.MenuValues
import java.util.Collections

class TimeSlider : ISlider, OnSliderTouchListener {
    private val activity: AppCompatActivity by lazy { MainActivity.getMainActivity() }
    private val slider: Slider = activity.findViewById(R.id.time_slider)
    private val minText: TextView = activity.findViewById(R.id.textMin)
    private val maxText: TextView = activity.findViewById(R.id.textMax)
    private val times: ArrayList<Int>

    init {
        slider.addOnSliderTouchListener(this)
        times = initTimes()
    }

    override fun show() {
        slider.visibility = View.VISIBLE
    }

    override fun hide() {
        slider.visibility = View.GONE
    }

    override fun initTimes(): ArrayList<Int> {
        val years = ArrayList<Int>()
        getYears(false).use { cursor ->
            while (!cursor.isAfterLast) {
                years.add(cursor.getInt(0))
                cursor.moveToNext()
            }
        }
        return years
    }

    override fun setTimesRange() {
        try {
            val minYear = Collections.min(times).toFloat()
            val maxYear = Collections.max(times).toFloat()
            slider.valueFrom = minYear
            slider.valueTo = maxYear
            slider.value = maxYear
            minText.text = minYear.toInt().toString()
            maxText.text = maxYear.toInt().toString()
        } catch (ex: Exception) {
            hideButton()
        }
    }

    override fun onStartTrackingTouch(slider: Slider) {
        minText.text = slider.value.toInt().toString()
    }

    override fun onStopTrackingTouch(slider: Slider) {
        minText.text = slider.value.toInt().toString()
        val min = slider.value.toString()
        val filter = "CAST(strftime('%Y',r.date) as int)==$min"
        AppPreferenceManager.setFilterSetter(MenuValues.SORT_DATE)
        AppPreferenceManager.setFilter(filter)
        refreshSelectedFragment()
    }
}
